### PR TITLE
Improve error when named argument is used with function pointer

### LINKED ISSUE #48071: Misleading diagnostic when using named arguments on function pointer call
**Version Used**: 6521491e7fbf9765797111faca482c9eb7f57116

**Steps to Reproduce**:

```cs
public class C {
    public unsafe void M(delegate*<string, int, void> ptr) {
        ptr("a", arg1: 1);
    }
}
```

[sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNGSAmIDUAfAAQGYACAgJhIGESBvAWACgSWzSBXAOyggDM4yAFhIBZABSY4AGzgBzCPABUAHgIBGAAxISAS04YhAPhIAHGAgCUdJq1unzYgEQRH2iAllqQJNRYDcNqwAvkxBQA)

**Expected Behavior**: An error is given indicating that function pointer invocations cannot have named arguments

**Actual Behavior**: `error CS8756: Function pointer 'delegate*<string, int, void>' does not take 2 arguments`

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #48071

@RikkiGibson Currently, the diagnostic is only produced for the first named argument only, let me know if you want to produce diagnostic for all named arguments and I'll give that a try.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 1]
```diff
@@ -476,7 +476,10 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
                     && !isMethodGroupConversion
                     // Function pointer type symbols don't have named parameters, so we just want to report a general mismatched parameter
                     // count instead of name errors.
-                    && !(firstSupported.Member is FunctionPointerMethodSymbol _))
+                    // The only case we want this block to be entered for function pointers is if it's called with a named argument.
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:479]: If function pointers are **always** allowed to enter this block, we will start to get strange error messages like:

> error CS7036: There is no argument given that corresponds to the required formal parameter '' of 'delegate*<string, void>'

which happens when on MemberResolutionKind.RequiredParameterMissing.

So my change only allowed function pointers to enter here for NoCorrespondingNamedParameter only.

**Comment by CyrusNajmabadi** [comment]
This seems strange to me.  Why do function pointer parameters even have names?  Either they have names (and then i would expect you can use named-params), or they should not have names (and then using named-params wouldn't even be possible).

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Function pointer parameters don't have names AFAIK, and using named arguments with them isn't allowed. But the current error message when using named arguments isn't clear.

**Comment by CyrusNajmabadi** [comment]
Gotcha.  the tests made me think there were actual named parameters like "arg0".

**Review by 333fred — APPROVED** [review]
LGTM (commit 4). @dotnet/roslyn-compiler for a second review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review of this community PR.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 2]
```diff
@@ -476,7 +476,10 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
...
+                    // We have a special error for it (ERR_FunctionPointersCannotBeCalledWithNamedArguments) reported
+                    // in ReportNoCorrespondingNamedParameter.
+                    && (firstSupported.Member is not FunctionPointerMethodSymbol || firstSupported.Result.Kind == MemberResolutionKind.NoCorrespondingNamedParameter))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:482]: >(firstSupported.Member is not FunctionPointerMethodSymbol || firstSupported.Result.Kind == MemberResolutionKind.NoCorrespondingNamedParameter) [](start = 23, length = 142)

I think the code would be easier to follow and maintain if we simply moved the check for function pointer and the resolution kind into a separate ```if``` branch and explicitly reported ```diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, location)``` from there without making any changes to ReportNoCorrespondingNamedParameter.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 3]
```diff
@@ -470,13 +470,22 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
...
+                    return;
+                }
                 // If there are multiple supported candidates, we don't have a good way to choose the best
                 // one so we report a general diagnostic (below).
-                if (!(firstSupported.Result.Kind == MemberResolutionKind.RequiredParameterMissing && supportedRequiredParameterMissingConflicts)
+                else if (!(firstSupported.Result.Kind == MemberResolutionKind.RequiredParameterMissing && supportedRequiredParameterMissingConflicts)
                     && !isMethodGroupConversion
                     // Function pointer type symbols don't have named parameters, so we just want to report a general mismatched parameter
                     // count instead of name errors.
-                    && !(firstSupported.Member is FunctionPointerMethodSymbol _))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:479]: >! [](start = 23, length = 1)

I think the negation should be kept.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:479]: Missed an addition of ```not```

---
In reply to: [496813646](https://github.com/dotnet/roslyn/pull/48081#discussion_r496813646) [](ancestors = 496813646)

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 4]
```diff
@@ -829,6 +838,13 @@ private void ReportDuplicateNamedArgument(MemberResolutionResult<TMember> result
 
             Location location = new SourceLocation(badName);
 
+            // Function pointers cannot be called with named arguments at all.
+            if (bad.Member is FunctionPointerMethodSymbol)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:842]: >if (bad.Member is FunctionPointerMethodSymbol) [](start = 12, length = 46)

Is this code still needed?

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 5]
```diff
@@ -470,13 +470,22 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
             // If there are any supported candidates, we don't care about unsupported candidates.
             if (firstSupported.IsNotNull)
             {
+                if (firstSupported.Member is FunctionPointerMethodSymbol
+                    && firstSupported.Result.Kind == MemberResolutionKind.NoCorrespondingNamedParameter)
+                {
+                    int badArg = firstSupported.Result.BadArgumentsOpt[0];
+                    IdentifierNameSyntax badName = arguments.Names[badArg];
+                    Location location = new SourceLocation(badName);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:478]: >location [](start = 29, length = 8)

It looks like this local conflicts with parameter

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5)

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs]: No, deleted.

**Comment by AlekseyTs** [comment]
It looks like there are legitimate test failures.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 6]
```diff
@@ -470,13 +470,21 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
             // If there are any supported candidates, we don't care about unsupported candidates.
             if (firstSupported.IsNotNull)
             {
+                if (firstSupported.Member is FunctionPointerMethodSymbol
+                    && firstSupported.Result.Kind == MemberResolutionKind.NoCorrespondingNamedParameter)
+                {
+                    int badArg = firstSupported.Result.BadArgumentsOpt[0];
+                    IdentifierNameSyntax badName = arguments.Names[badArg];
+                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, location);
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:478]:
```suggestion
                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, new SourceLocation(badName));
```

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 7]
```diff
@@ -475,7 +475,7 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
                 {
                     int badArg = firstSupported.Result.BadArgumentsOpt[0];
                     IdentifierNameSyntax badName = arguments.Names[badArg];
-                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, location);
+                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, new SourceLocation(badName));
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:478]: `badName.Location`

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:478]: The pattern `new SourceLocation(badName)` is used multiple times in existing code in this file. Do you want to update all of them or just this new code?

I also don't see a Location property on IdentifierNameSyntax through the API doc:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.syntax.identifiernamesyntax?view=roslyn-dotnet#properties

So I'll use `GetLocation()` method.

```suggestion
                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, badName.GetLocation());
```

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs]: All `SyntaxNode`s have an internal `Location` API. Please use that: http://sourceroslyn.io/#Microsoft.CodeAnalysis/Syntax/SyntaxNode.cs,f38d78484e47462d.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs [hunk 8]
```diff
@@ -470,13 +470,21 @@ private static ThreeState TryGetBestResult(ArrayBuilder<MemberResolutionResult<T
             // If there are any supported candidates, we don't care about unsupported candidates.
             if (firstSupported.IsNotNull)
             {
+                if (firstSupported.Member is FunctionPointerMethodSymbol
+                    && firstSupported.Result.Kind == MemberResolutionKind.NoCorrespondingNamedParameter)
+                {
+                    int badArg = firstSupported.Result.BadArgumentsOpt[0];
+                    IdentifierNameSyntax badName = arguments.Names[badArg];
+                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, badName.GetLocation());
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:478]:
```suggestion
                    diagnostics.Add(ErrorCode.ERR_FunctionPointersCannotBeCalledWithNamedArguments, badName.Location);
```

**Review by AlekseyTs — APPROVED** [review]
LGTM (itertaion 9)

**Comment by AlekseyTs** [comment]
@333fred Please review the latest revision.

**Review by 333fred — APPROVED** [review]
LGTM (commit 9)

**Comment by 333fred** [comment]
Thanks Youssef!

**Pull request merged**